ledger/common/hash.go

```diff
@@ -38,33 +39,26 @@ func GetDefaultHashForHeight(height int) []byte {
 }
 
 // HashLeaf generates hash value for leaf nodes (SHA3-256).
+//
+// path must be a 32 byte slice.
 // note that we don't include the keys here as they are already included in the path
 func HashLeaf(path []byte, value []byte) []byte {
-	hasher := hash.NewSHA3_256()
-	_, err := hasher.Write(path)
-	if err != nil {
-		panic(err)
-	}
-	_, err = hasher.Write(value)
-	if err != nil {
-		panic(err)
-	}
-
-	return hasher.SumHash()
+	hasher := new256()
+	hasher.write256(path) // path is always 256 bits
+	hasher.write(value)
+	hasher.finalize()
+	return hasher.hashResult()
 }
 
 // HashInterNode generates hash value for intermediate nodes (SHA3-256).
+//
+// hash1 and hash2 must each be a 32 byte slice.
 func HashInterNode(hash1 []byte, hash2 []byte) []byte {
-	hasher := hash.NewSHA3_256()
-	_, err := hasher.Write(hash1)
-	if err != nil {
-		panic(err)
-	}
-	_, err = hasher.Write(hash2)
-	if err != nil {
-		panic(err)
-	}
-	return hasher.SumHash()
+	hasher := new256()
+	hasher.write256(hash1) // hash1 and hash2 are 256 bits
+	hasher.write256(hash2)
+	hasher.finalize512()
```

Comment:
Thanks @AlexHentschel. I agree, I think this kind of optimizations will have a good impact. In this PR, I only focused on optimizing the sha3 function itself, independently of the trie module. The next issue is the one looking at how hashing is called in the merkle trie, where ideas like what you described can be used.

I also wanted to clarify a point about this current PR: 
 - There used to be a dynamic allocation of a slice to store the hash result, and the timing improved by deleting it.
 - The result now is stored in the hasher structure itself. `buf` [here](https://github.com/onflow/flow-go/blob/703a695a04bee98bea12ae2e264ee5d3498836a3/ledger/common/sha3.go#L73) is a slice that is never allocated, it is just a pointer [to another array](https://github.com/onflow/flow-go/blob/703a695a04bee98bea12ae2e264ee5d3498836a3/ledger/common/sha3.go#L76) of constant size, within the structure itself.
  - This means the entire hashing structure `state` is declared locally in sha3, but a slice pointing to an internal array is what is returned as a result. I'm not sure how the GC is handling all this, but it somehow gives better results than dynamic allocations. 
  - Ideally, I would like to have the result array being passed as input to sha3, and make the trie algos (calling sha3) use a minimum number of arrays.

Reply:
> Ideally, I would like to have the result array being passed as input to sha3, and make the trie algos (calling sha3) use a minimum number of arrays.

this was essentially at the core of my suggestion.

Reply:
Thanks for the clarifications on the implementation. Now that I understand a bit better what is going on with  `buf` and `storage`, I have a few thoughts regarding API safety and memory footprint: 

#### API safety 
How do you make sure that you never exceed the capacity of 
https://github.com/onflow/flow-go/blob/703a695a04bee98bea12ae2e264ee5d3498836a3/ledger/common/sha3.go#L215-L218
* HashInterNode accepts variable-length byte slices: https://github.com/onflow/flow-go/blob/703a695a04bee98bea12ae2e264ee5d3498836a3/ledger/common/hash.go#L56 
* I understand that the _current_ implementation does not exceed the capacity of `storage` ... but it is super easy to break in future refactorings. 

#### memory footprint
* a hash length is 32 bytes
* but each byte slice with the result is backed by an array of size `maxRate = 168` https://github.com/onflow/flow-go/blob/703a695a04bee98bea12ae2e264ee5d3498836a3/ledger/common/sha3.go#L97 Hence, your hashes will have a memory footprint that is **5 times larger** than needed

Reply:
> How do you make sure that you never exceed the capacity of the array that is backing buf ?

The algorithm and implementation assures it by construction. sha3 "absorbs" the input chunk by chunk, so that the internal state has a constant size. That can be seen in the code too. 

> if we ever exceed it, the hash will be badly broken, right?

Yes it would panic. 

> HashInterNode accepts variable-length byte slices

The function `HashInterNode` ONLY works with 2 inputs of 32 bytes each. It's not a generic one. Similarly, `HashLeaf` ONLY work when the first input is 32 bytes (no restriction on the second input). If these requirements change in the future, we need to write new functions.
The caller MUST meet these conditions, lengths are not checked within the functions to save the checking time.

> Hence, your hashes will have a memory footprint that is 5 times large

Yes that's correct, happy you pointed this out, and this is why I said I don't know how the GC will handle this :) Since the output is a slice (with length and capacity) and not the entire array, I am wondering what will happen: keep the entire array in memory or only keep the portion of the slice capacity. Even worse, it may keep the entire `state` structure.
I've tried a couple things to avoid this issue (breaking `storage` into 32 bytes and the rest, for instance) but no success for now. 
If you think EN memory won't support this, I could add a final extra copy of the 32 bytes of result, but this will definitely slow the function down.

Btw, I've updated [the storage size](https://github.com/onflow/flow-go/commit/b80b1cd93afe481a381a90cf1230d953eb075432) from 168 to 136. 168 is not needed for sha3-256.